a8c-files.php

```diff
@@ -776,6 +776,232 @@ function a8c_files_init() {
+	 * @param array  $size_array    Array of width and height values in pixels (in that order).
+	 * @param string $image_src     The 'src' of the image.
+	 * @param int    $attachment_id The image attachment ID or 0 if not supplied.
+	 * @return array                The normalized image meta data or empty $image_meta['sizes'] on error.
+	 */
+	public static function generate_image_meta( $image_meta, $size_array, $image_src, $attachment_id ) {
+		// Reset $image_meta['sizes'] if not empty. Need to build it again matching the current img src.
+		$image_meta['sizes'] = array();
+
+		// Bail out early if the image is not hosted on the site, missing image meta or original image is too small.
+		// This will disable the generation of srcset.
+		if ( strpos( $image_src, home_url() ) === false ||
+			empty( $image_meta['width'] ) ||
+			empty( $image_meta['height'] ) ||
+			( $image_meta['width'] < 100 && $image_meta['height'] < 100 ) ) {
+
+			return $image_meta;
+		}
+
+		$image_basename = wp_basename( $image_meta['file'] );
+
+		// Return early if filename mismatch
+		if ( strpos( wp_basename( $image_src ), $image_basename ) === false ) {
+			return $image_meta;
+		}
+
+		// Need the mime type but without getting the attachment post (fast).
+		if ( ! preg_match( '/\.(jpe?g|png|gif)$/i', $image_basename, $extension ) ) {
+			// Something doesn't add up...
+			return $image_meta;
+		}
+
+		$extension = strtolower( $extension[1] );
+
+		if ( 'jpg' === $extension || 'jpeg' === $extension ) {
+			$mime_type = 'image/jpeg';
+		} else {
+			$mime_type = "image/{$extension}";
+		}
+
+		$crop = false;
+
+		// Add the Photon query args from the img src.
+		// Also, when adding the (new) sizes, one of the 'file' values has to match the image src basename
+		// including the query args. Otherwise wp_calculate_image_srcset() bails out and returns false.
+		if ( strpos( $image_src, '?' ) ) {
+			$query = explode( '?', $image_src );
+			wp_parse_str( str_replace( array( '&#38;', '&#038;', '&amp;' ), '&', $query[1] ), $photon_args );
+
+			if ( isset( $photon_args['zoom'] ) ) {
+				// This should only be used by direct requests from JS/devicepx.
+				// No good way to generate proper srcset for "zoomed" images.
+				return $image_meta;
+			}
+
+			if ( isset( $photon_args['fit'] ) ) {
+				// When there are width and height from the img tag (that translate into "w" and "h")
+				// the "fit" arg is not needed. If used, it is probably a "special case" or an error.
+				return $image_meta;
+			}
+
+			if ( isset( $photon_args['crop'] ) ) {
+				if ( $photon_args['crop'] === '1' ) {
+					$crop = true;
+				} else {
+					// No support for Photon cropping using pixel values or percents, yet.
+					return $image_meta;
+				}
+			}
+
+			if ( ! empty( $photon_args['resize'] ) ) {
+				// TODO: maybe see if there is a discrepancy between the width/height from the img tag and the values in 'resize'
```

Comment:
Probably best to just remove it, ideally we keep aligned with the dotcom version going forward.